File: src/InferenceEngine/python/interface/mqttproducer.py
import paho.mqtt.client as mqtt
import json

class MQTTProducer:

    def __init__(self, server, client_id):
        self.server = server
        self.client_id = client_id
        self.client = mqtt.Client(self.client_id)
    
    def __del__(self):
        self.stop()
    
    def start(self):
        self.client.connect(self.server["address"], self.server["port"])

    def stop(self):
        self.client.disconnect()

    def produce(self, topic, msg):
        msg_str = json.dumps(msg)
        self.client.publish(topic, msg_str)

File: src/InferenceEngine/cpp/interface/mqttproducer.cpp
#include "mqttproducer.hpp"

MQTTProducer::MQTTProducer(const std::string& server, const std::string& clientID)
    : serverAddress(server), clientId(clientID), client(serverAddress, clientId, mqtt::create_options(MQTTVERSION_5)) {}

MQTTProducer::~MQTTProducer() { this->stop(); }

void MQTTProducer::start() { this->client.connect(); }

void MQTTProducer::stop() { this->client.disconnect(); }

void MQTTProducer::produce(const std::string& topic, const nlohmann::json& msg) {

    std::string msgStr = msg.dump();
    mqtt::message_ptr msgPointer = mqtt::make_message(topic, msgStr);
    msgPointer->set_payload(msgStr);
    this->client.publish(msgPointer);

}